# Introductory Notes

Throughout this entire notebook you should be experimenting with the code in the non-text cells. A great way to begin to get a feel for Python is by playing with it. So have some fun by changing the values in the cells and then running them again with Shift-Enter. Before you do, think about what you expect the output to be, and make sure your intuition matches up with what you run. If it doesn't, take some time to think about what happened so you can hone your intuition.

At the end of each section there will be some questions to help further your understanding. Remember, in Python we can always manually test code by running it; however, you should try to think about the answers to these questions before you run some code. This way you can check and verify your understanding of the section's topic.

### List and Dictionary Comprehensions

The last topic we'll cover today is a different way to construct lists and dictionaries. Up to now, every time that we have built up a list or dictionary, we began by initializing it. We then took advantage of their mutability inherent to build them up one element or key-value pair at a time. However, there is a more succinct way to accomplish the vast majority of your list and dictionary construction tasks.

#### List Comprehensions

Before we dive into the specifics about how this new tool (list comprehensions) works, let's look at an example question where we build a list.  We can then show how to perform the same task with our new tool and learn how it works.

Let's imagine that we have the list `[1, 5, 9, 33]` stored in the variable `my_list`. Now, let's assume that we want to make a new list of the squares of all the values in `my_list` and call it `my_squares`. With the tools we have covered so far, you might write:

In [1]:
my_list = [1, 5, 9, 33]
my_squares = []
for num in my_list:
    my_squares.append(num ** 2)
print(my_squares)

[1, 25, 81, 1089]


Now, `my_squares` will hold the list `[1, 25, 81, 1089]`. To get this, we were simply specifying a bunch of stuff that we wanted to add on to the end of the `my_squares` list, with a starting point at `my_list`. So, from a high level, we can write the framework of creating a list in code as:

```python
list_were_building = []
for thing in iterable:
    list_were_building.append(transform(thing))
```

With this structure in mind, we can use the following syntax to perform the same task of building up a list in a single line! Check it out, along with how it would look for the construction of `my_squares`.

```python
list_were_building = [transform(thing) for thing in iterable]
```

This last line of code does the exact same thing as the three lines above! In this line, the thing that we would pass to the `append()` method, `transform(thing)`, comes at the beginning of the statement in the `[]`.  These `[]` allow for the final product to be defined as a list. Then, the `for` loop statement that we had written is at the end. This is the basic idea behind the [list comprehension](https://en.wikipedia.org/wiki/List_comprehension).

Similarly, we can build our `my_squares` list using a list comprehension:

In [2]:
my_squares2 = [num ** 2 for num in my_list]
print(my_squares2)

[1, 25, 81, 1089]


But wait! There's more! Remember in all the examples where we were getting evens, we had a condition to decide when to append a value to a list? We can also use conditions to determine what "transformed things" get added in a list comprehension! Let's look at the evens list builder to hammer this home.

```python
# Old way of constructing list of evens
evens = []
for num in range(10):
    if num % 2 == 0:
        evens.append(num)

# Old way at high level
list_were_building = []
for thing in iterable:
    if condition:
        list_were_building.append(transform(thing))

# List comprehension way of constructing list of evens
evens = [num for num in range(10) if num % 2 == 0]

# List comprehension way at high level
list_were_building = [transform(thing) for thing in iterable if condition]
```
The way `transform()` was called in the above examples, as though it were a function, is an option when writing list comps. For example, the `my_squares` example could be accomplished in the same way with:

```python
def square(num):
    return num ** 2

my_squares = [square(num) for num in my_list]
```

This might seem silly, since we could just write `num ** 2` directly in the list comp as we did above. However, this calling of a function in the list comp becomes a powerful idea when you want to transform the values being iterated over in a complex way.

#### Dictionary Comprehensions

Just as list comprehensions are a more succinct way of constructing a list, we have the same ability for dictionaries. Dictionary comprehensions operate in the same way as their list counterparts, except for one fundamental difference. Recall that dictionaries have no `append()` method, and that a new key-value pair is added to the dictionary with the syntax: `my_dict[new_key] = new_value`. In this way, it makes sense that we need syntax to pass both the key and value to the dictionary comprehension.

Luckily, Python gives a simple way to pass a key and value pair, and it is already very familiar to you! You just separate the key and value that you want to enter into the dictionary with a colon, like we did when we were hardcoding the contents in the `{}` dictionary constructor, i.e. `my_dict = {1: 1, 2: 4}`. Let's look at an example where we make a dictionary with the keys as the numbers 1 - 5, and the values as the squares of the keys. We'll do this with both the old way of constructing a dictionary, and then with a dictionary comprehension so that we can see the similarities.

In [3]:
# Standard way. 
squares_dict = {}
for num in range(1, 6): 
    squares_dict[num] = num ** 2
print squares_dict

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


In [4]:
# Dictionary Comprehension way. 
squares_dict2 = {num: num ** 2 for num in range(1, 6)}
print squares_dict2

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


We can see that in both cases, we're going through the numbers 1 - 5 with `range(1, 6)` and those `num`s are being assigned as keys. The values assigned to those keys are the squares of the keys, assigned with `squares_dict[num] = num ** 2` and `num: num ** 2`, respectively. Just as with list comprehensions, dictionary comprehensions read as the first thing being the `key: value` pair being added to the dictionary. Then, left to right (top down in the old way), we have what the loop definition would look like. And, just as with list comps, we can add a condition to filter what gets put into the dictionary.

Say that we want a dictionary with a random integer between 1 and 10, associated with each of the values in the list of words: `['cow', 'chicken', 'horse', 'moose']`. Let's look at how we'd do that with a dictionary comprehension. (We're importing from the Python library `random` to get our random integers. We'll talk more about importing later in the course.)

In [5]:
from random import randint
animals_list = ['cow', 'chicken', 'horse', 'moose']

In [6]:
animals_dict = {animal: randint(1, 10) for animal in animals_list}
print animals_dict

{'moose': 9, 'chicken': 8, 'horse': 6, 'cow': 6}


#### Other Comprehensions

You can actually use the syntax from the list comprehensions to construct a tuple in what seems like a dynamic way. Take the example.


In [7]:
my_tuple = tuple(num for num in range(10) if num % 2 == 0)
print my_tuple

(0, 2, 4, 6, 8)


All we are doing here is passing `num for num in range(10) if num % 2 == 0` to the tuple constructor, `()`. Since the tuple constructor takes any iterable, which that statement produces, it makes a tuple out of the contents. Note that it would be impossible to make a tuple with statements like this the "old way", since tuples don't support appending or mutation of any kind!

For this reason, in addition to their readability, comprehensions of all types are considered the most Pythonic way of constructing new data structures.

##### Construction Questions 

1. Take the following for loop, and translate it into a list. comprehensions: 
      ```python 
           odds = []
           for num in range(10): 
               if num % 2 != 0: 
                   odds.append(num)
       ```
       
1. Take the following for loop, and translate it into a dictionary comprehensions: 
      ```python 
           cubes = {}
           for num in range(1, 6): 
               cubes[num] = num **  3
       ```